geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java

```diff
@@ -2335,6 +2330,16 @@ public void membershipFailure(String reason, Throwable t) {
       }
     }
 
+    @Override
+    public Throwable setShutdownCause(Throwable cause) {
+      if (cause != null && !(cause instanceof ForcedDisconnectException)) {
```

Comment:
Reinterpreting the `cause` sent to this listener method breaks the contract. The listener is a recipient of information (in this case, from membership). What if the reason for shutdown was not a forced disconnect?

Also wrong, is calling this listener method from the `membershipFailure()` method above. This method is a callback that should only be invoked by membership.

In fact before this method is invoked from `membershipFailure()` it has already been invoked by membership.

Reply:
Here only refactored the existing code into a method, without changing any logic. 

See the replaced code, it changed cause to be ForcedDisconnectException.